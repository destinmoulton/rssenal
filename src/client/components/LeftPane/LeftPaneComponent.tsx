import { OrderedMap } from "immutable";
import * as React from "react";

import { Icon } from "semantic-ui-react";

import AddFeedModalContainer from "../../containers/Modals/AddFeedModalContainer";
import ButtonBarContainer from "../../containers/LeftPane/ButtonBarContainer";
import EditFeedModalContainer from "../../containers/Modals/EditFeedModalContainer";
import EditFolderModalContainer from "../../containers/Modals/EditFolderModalContainer";
import FolderItemContainer from "../../containers/LeftPane/FolderItemContainer";

import { propertyComparator } from "../../lib/sort";

import { TFeeds, IFeed, IFolder, TFeedID, TFolderID } from "../../types";

export interface ILeftPaneMapState {
    feeds: TFeeds;
    hasFolders: boolean;
    folders: OrderedMap<TFolderID, IFolder>;
}

export interface ILeftPaneMapDispatch {
    getAllFolders: () => void;
    getAllFeeds: () => void;
}

type IAllProps = ILeftPaneMapDispatch & ILeftPaneMapState;

interface ILeftPaneState {
    addFeedModalOpen: boolean;
    editFeed: IFeed;
    editFeedModalOpen: boolean;
    editFolder: IFolder;
    editFolderModalOpen: boolean;
}
class LeftPaneComponent extends React.Component<IAllProps, ILeftPaneState> {
    state: ILeftPaneState = {
        addFeedModalOpen: false,
        editFeed: null,
        editFeedModalOpen: false,
        editFolder: null,
        editFolderModalOpen: false
    };

    componentDidMount() {
        const { getAllFolders, getAllFeeds, hasFolders } = this.props;

        if (!hasFolders) {
            getAllFolders();
            getAllFeeds();
        }
    }

    _handleOpenAddFeedModal = (feed: IFeed) => {
        this.setState({
            editFeed: feed,
            addFeedModalOpen: true
        });
    };

    _handleCloseAddFeedModal = () => {
        this.setState({
            addFeedModalOpen: false
        });
    };

    _handleOpenEditFolderModal = (folder: IFolder) => {
        this.setState({
            editFolder: folder,
            editFolderModalOpen: true
        });
    };

    _handleCloseEditFolderModal = () => {
        this.setState({
            editFolder: null,
            editFolderModalOpen: false
        });
    };

    _handleOpenEditFeedModal = (feed: IFeed) => {
        this.setState({
            editFeed: feed,
            editFeedModalOpen: true
        });
    };

    _handleCloseEditFeedModal = () => {
        this.setState({
            editFeed: null,
            editFeedModalOpen: false
        });
    };

    render() {
        const {
            addFeedModalOpen,
            editFeed,
            editFeedModalOpen,
            editFolder,
            editFolderModalOpen
        } = this.state;

        const { folders } = this.props;
        const sortedFolders = folders
            .sort((a: IFolder, b: IFolder) =>
                propertyComparator<IFolder, number>(a, b, "asc", "order", false)
            )
            .toArray();
        sortedFolders.unshift({ name: "All", _id: "all", order: -1 });

        let listFolders = sortedFolders.map(folder => {
            return (
                <div key={folder._id}>
                    <FolderItemContainer
                        folder={folder}
                        editFolder={this._handleOpenEditFolderModal}
                        editFeed={this._handleOpenEditFeedModal}
                    />
                </div>
            );
        });

        return (
            <div>
                <div>
                    <ButtonBarContainer
                        openEditFolderModal={this._handleOpenEditFolderModal}
                        openAddFeedModal={this._handleOpenAddFeedModal}
                    />
                    <AddFeedModalContainer
                        isModalOpen={addFeedModalOpen}
                        onCloseModal={this._handleCloseAddFeedModal}
                    />
                    <EditFolderModalContainer
                        isModalOpen={editFolderModalOpen}
                        onCloseModal={this._handleCloseEditFolderModal}
                        folder={editFolder}
                    />
                    <EditFeedModalContainer
                        isModalOpen={editFeedModalOpen}
                        onCloseModal={this._handleCloseEditFeedModal}
                        feed={editFeed}
                    />
                </div>
                <div className="rss-leftmenu-folders-box">{listFolders}</div>
                <div className="rss-leftmenu-logo">
                    <Icon name="rss" />&nbsp;RSSenal
                </div>
            </div>
        );
    }
}

export default LeftPaneComponent;
