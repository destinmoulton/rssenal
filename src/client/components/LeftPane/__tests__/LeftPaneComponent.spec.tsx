import { OrderedMap } from "immutable";
import * as React from "react";

import { shallow } from "enzyme";

import LeftPaneComponent from "../LeftPaneComponent";
import feeds from "../../../../../test/data/immutable/feeds";
import folders from "../../../../../test/data/immutable/folders";

describe("<LeftPaneComponent />", () => {
    describe("renders and matches the snapshot", () => {
        it("matches snapshot when hasFolders = false", () => {
            const wrapper = shallow(
                <LeftPaneComponent
                    foldersGetAll={jest.fn()}
                    feedsGetAll={jest.fn()}
                    folders={OrderedMap()}
                    feeds={feeds}
                    hasFolders={false}
                />
            );
            expect(wrapper).toMatchSnapshot();
        });
        it("matches snapshot when hasFolders = true", () => {
            const wrapper = shallow(
                <LeftPaneComponent
                    foldersGetAll={jest.fn()}
                    feedsGetAll={jest.fn()}
                    folders={folders}
                    feeds={feeds}
                    hasFolders={true}
                />
            );
            expect(wrapper).toMatchSnapshot();
        });
    });
});
