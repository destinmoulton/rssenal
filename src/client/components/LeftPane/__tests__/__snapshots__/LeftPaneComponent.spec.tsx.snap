// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LeftPaneComponent /> renders and matches the snapshot matches snapshot when hasFolders = false 1`] = `
<div>
  <div>
    <Connect(ButtonBarComponent)
      openAddFeedModal={[Function]}
      openEditFolderModal={[Function]}
    />
    <Connect(AddFeedModalComponent)
      isModalOpen={false}
      onCloseModal={[Function]}
    />
    <Connect(EditFolderModalComponent)
      folder={null}
      isModalOpen={false}
      onCloseModal={[Function]}
    />
    <Connect(EditFeedModalComponent)
      feed={null}
      isModalOpen={false}
      onCloseModal={[Function]}
    />
  </div>
  <div
    className="rss-leftmenu-folders-box"
  >
    <div
      key="all"
    >
      <Connect(FolderItemComponent)
        editFeed={[Function]}
        editFolder={[Function]}
        folder={
          Object {
            "_id": "all",
            "name": "All",
            "order": -1,
          }
        }
      />
    </div>
  </div>
  <div
    className="rss-leftmenu-logo"
  >
    <Icon
      as="i"
      name="rss"
    />
     RSSenal
  </div>
</div>
`;

exports[`<LeftPaneComponent /> renders and matches the snapshot matches snapshot when hasFolders = true 1`] = `
<div>
  <div>
    <Connect(ButtonBarComponent)
      openAddFeedModal={[Function]}
      openEditFolderModal={[Function]}
    />
    <Connect(AddFeedModalComponent)
      isModalOpen={false}
      onCloseModal={[Function]}
    />
    <Connect(EditFolderModalComponent)
      folder={null}
      isModalOpen={false}
      onCloseModal={[Function]}
    />
    <Connect(EditFeedModalComponent)
      feed={null}
      isModalOpen={false}
      onCloseModal={[Function]}
    />
  </div>
  <div
    className="rss-leftmenu-folders-box"
  >
    <div
      key="all"
    >
      <Connect(FolderItemComponent)
        editFeed={[Function]}
        editFolder={[Function]}
        folder={
          Object {
            "_id": "all",
            "name": "All",
            "order": -1,
          }
        }
      />
    </div>
    <div
      key="5a54f74838e78863ac6a8961"
    >
      <Connect(FolderItemComponent)
        editFeed={[Function]}
        editFolder={[Function]}
        folder={
          Object {
            "__v": 0,
            "_id": "5a54f74838e78863ac6a8961",
            "name": "Lifestyle",
            "order": 1,
          }
        }
      />
    </div>
    <div
      key="5a53ec3a94c3905e4a8328d4"
    >
      <Connect(FolderItemComponent)
        editFeed={[Function]}
        editFolder={[Function]}
        folder={
          Object {
            "__v": 0,
            "_id": "5a53ec3a94c3905e4a8328d4",
            "name": "Programming",
            "order": 2,
          }
        }
      />
    </div>
    <div
      key="5a54fcf538e78863ac6a89f7"
    >
      <Connect(FolderItemComponent)
        editFeed={[Function]}
        editFolder={[Function]}
        folder={
          Object {
            "__v": 0,
            "_id": "5a54fcf538e78863ac6a89f7",
            "name": "Writing",
            "order": 3,
          }
        }
      />
    </div>
    <div
      key="5a55035e38e78863ac6a89fe"
    >
      <Connect(FolderItemComponent)
        editFeed={[Function]}
        editFolder={[Function]}
        folder={
          Object {
            "__v": 0,
            "_id": "5a55035e38e78863ac6a89fe",
            "name": "Random",
            "order": 4,
          }
        }
      />
    </div>
  </div>
  <div
    className="rss-leftmenu-logo"
  >
    <Icon
      as="i"
      name="rss"
    />
     RSSenal
  </div>
</div>
`;
